### PR TITLE
[WIP; RFC; Pending 1268, Stacked on 1270] Enhancement: git tag sem-ver resolution

### LINKED ISSUE #925: git dependencies should allow semver requirement
- [x] I have searched the [issues](https://github.com/sdispater/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have searched the [documentation](https://poetry.eustace.io/docs/) and believe that my question is not covered.

## Feature Request

First thanks for creating this much needed tool for the python community.

I don't see this in the documentation or an existing issue for it, but it would be awesome if I could have a git dependency something like this:
```
requests = { git = "https://github.com/kennethreitz/requests.git", semver = "^2.20.0" }
```

That would allow for automatic updating based on the git tags that meet the semver requirement and match this regex in their tag name: `^v?\d+\.\d+\.\d+$` (or something similar).

As an example, NPM allows for this https://docs.npmjs.com/files/package.json#git-urls-as-dependencies

### PR DESCRIPTION
**Pull request by darrenleeweber**
# Issue(s)

- fix #925

# Split out PRs

Not necessarily stacked, but isolated changes in smaller PRs:
- [x] #1266 - optional 
- [x] #1267 - optional
- [ ] #1268 - required
- [ ] #1269 - required (although sorta optional refactor)
- [ ] #1270 - required, stacked on #1269 
- [ ] #1176 now stacked on #1270
  - add sem-ver git tags and provider support
  - on-hold while PRs above are processed
    - minor cleanup and updates required

# Check List

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.
- [x] Rebase on develop for new features and change PR target branch to develop

### CODE REVIEW

File: tests/puzzle/test_provider.py [hunk 1]
```diff
@@ -95,6 +95,63 @@ def test_search_for_vcs_read_setup(provider, mocker):
...
+    assert dependency.branch is None
+    assert dependency.rev is None
+    assert dependency.tag == "0.2.0"
+
+    packages = provider.search_for_vcs(dependency)
+    assert len(packages) == 1
+    package = packages[0]
+    assert package.name == "demo"
+    assert package.version.text != dependency.tag
+    assert package.version.text == "0.1.2"
```
Comment (darrenleeweber) [inline tests/puzzle/test_provider.py]: This is the current behavior, captured in a test.  Not clear yet how/why it gets "0.1.2" here (must be in mocks somewhere).

**Comment by darrenleeweber** [comment]
tests are failing because test fixtures for a git repo do not support a `git tag -l` command

**Comment by darrenleeweber** [comment]
Splitting this PR out into smaller PRs

**Comment by sdispater** [comment]
I don't think that's something I am willing to add to Poetry. This is a very specific workflow (even though I understand why you might want it) that does not make sense to add to the Poetry code base.

**Comment by darrenleeweber** [comment]
Interesting opinion, thanks for the transparency on that, it could save a lot of work.  The options then are:
- fork permanently, which is not attractive
- a plugin solution to provide this option
  - how does/could that work?

**Comment by sdispater** [comment]
@darrenleeweber The truth is, that's a workflow that I was not aware of until now and npm seems to be the only one to support this.

But yes, going the plugin route might be the preferred way to implement this once the plugin system is done (see #1237)

**Pull request closed without merging**